### PR TITLE
[RFC] Detect system default resolver

### PR DESCRIPTION
**Pull request by WyriHaximus**
As discussed in #29 a PR that detects the system default resolver on *nix systems.

### CODE REVIEW

**Comment by jsor** [comment]
#48 adds a TimeoutExecutor, would it make sense to read the [timeout option](https://wiki.archlinux.org/index.php/Resolv.conf#Use_timeout_option_to_reduce_hostname_lookup_time) from resolve.conf and use that value to create the TimeoutExecutor?

**Comment by WyriHaximus** [comment]
> #48 adds a TimeoutExecutor, would it make sense to read the timeout option from resolve.conf and use that value to create the TimeoutExecutor?

Most certainly! If @reactphp/core likes the proposed external API I'll go more in depth and add more features like the timeout option :smile: .

**Comment by mailopl** [comment]
Any news on this guys? Any chance for it to being merged in next week?

**Comment by kelunik** [comment]
What's with Windows? This PR will result in many written apps not being able to run on Windows.

**Comment by WyriHaximus** [comment]
Ping @jsor @clue, what do you think about the public facing API? If you like it I'll work it out into something that also works on windows and has a default fallback so it works if detection fails.

**Comment by WyriHaximus** [comment]
@clue we could go with something like this, where `8.8.8.8` is used as a fallback:

```php
$loop = React\EventLoop\Factory::create();
$factory = new React\Dns\Resolver\Factory();
$dns = $factory->createCached(determineSystemDefaultServer('8.8.8.8'), $loop);
```

File: src/Resolver/Factory.php [hunk 1]
```diff
@@ -62,4 +74,27 @@ protected function addPortToServerIfMissing($nameserver)
...
+                $nameserverPosition = stripos($line, 'nameserver');
+                if ($nameserverPosition !== false) {
+                    $nameserverLine = trim(substr($line, $nameserverPosition + 11));
+                    list ($nameservers[]) = explode(' ', $nameserverLine);
+                }
+            }
+        }
+
+        if (count($nameservers) > 0) {
+            shuffle($nameservers);
```
Comment (kelunik) [inline src/Resolver/Factory.php]: I think it should use the first one, not a random one. I think the first one is considered the primary DNS and order should be respected in `/etc/resolv.conf`.

File: src/Resolver/Factory.php [hunk 2]
```diff
@@ -62,4 +74,27 @@ protected function addPortToServerIfMissing($nameserver)
...
+    {
+        $nameservers = array();
+
+        if (stripos(PHP_OS, 'win') === false) {
+            $lines = file(self::UNIX_RESOLVE_CONF);
+            foreach ($lines as $line) {
+                $nameserverPosition = stripos($line, 'nameserver');
+                if ($nameserverPosition !== false) {
+                    $nameserverLine = trim(substr($line, $nameserverPosition + 11));
+                    list ($nameservers[]) = explode(' ', $nameserverLine);
```
Comment (kelunik) [inline src/Resolver/Factory.php]: There can be only one nameserver per `nameserver` config option in `/etc/resolv.conf`.

File: src/Resolver/Factory.php [hunk 3]
```diff
@@ -62,4 +74,27 @@ protected function addPortToServerIfMissing($nameserver)
...
+                }
+            }
+        }
+
+        if (count($nameservers) > 0) {
+            shuffle($nameservers);
+            return array_pop($nameservers);
+        }
+
+        throw new \Exception('Nameserver configuration missing');
```
Comment (kelunik) [inline src/Resolver/Factory.php]: `man resolv.conf` says it should default to the local nameserver if none are present, but Amp doesn't do this either, at least currently.

**Comment by kelunik** [comment]
I added some comments regarding the implementation.

While it will usually work to read `/etc/resolv.conf` once, the nameservers in there might change, e.g. because a software might be running on a notebook that changes the access point (WLAN).

**Pull request closed without merging**